kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandEncryptionManager.java

```diff
@@ -0,0 +1,263 @@
+        if (attempt >= MAX_ATTEMPTS) {
+            return CompletableFuture.failedFuture(
+                    new RequestNotSatisfiable("failed to reserve an EDEK to encrypt " + allRecordsCount + " records for topic " + topicName + " partition "
+                            + partition + " after " + attempt + " attempts"));
+        }
+        return currentDek(encryptionScheme).thenCompose(dek -> {
+            // if it's not alive we know a previous encrypt call has removed this stage from the cache and fall through to retry encrypt
+            if (!dek.isDestroyed()) {
+                try (Dek<E>.Encryptor encryptor = dek.encryptor(allRecordsCount)) {
+                    BatchAwareMemoryRecordsBuilder encrypt = encryptBatches(encryptionScheme, records, encryptor, bufferAllocator);
+                    return CompletableFuture.completedFuture(encrypt);
+                }
+                catch (ExhaustedDekException e) {
+                    rotateKeyContext(encryptionScheme, dek);
+                    // fall through to recursive call below...
+                }
+                catch (Exception e) {
+                    return CompletableFuture.failedFuture(e);
+                }
+            }
+            // recurse, incrementing the attempt number
+            return attemptEncrypt(topicName, partition, encryptionScheme, records, attempt + 1, batchRecordCounts, bufferAllocator);
+        });
+    }
+
+    @NonNull
+    private BatchAwareMemoryRecordsBuilder encryptBatches(@NonNull EncryptionScheme<K> encryptionScheme,
+                                                          @NonNull MemoryRecords memoryRecords,
+                                                          @NonNull Dek.Encryptor encryptor,
+                                                          @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator) {
+        BatchAwareMemoryRecordsBuilder builder = new BatchAwareMemoryRecordsBuilder(allocateBufferForEncrypt(memoryRecords, bufferAllocator));
+        for (MutableRecordBatch batch : memoryRecords.batches()) {
+            maybeEncryptBatch(encryptionScheme, encryptor, batch, builder);
+        }
+        return builder;
+    }
+
+    private void maybeEncryptBatch(@NonNull EncryptionScheme<K> encryptionScheme,
+                                   @NonNull Dek.Encryptor encryptor,
+                                   @NonNull MutableRecordBatch batch,
+                                   @NonNull BatchAwareMemoryRecordsBuilder builder) {
+        if (batch.isControlBatch()) {
+            builder.writeBatch(batch);
+        }
+        else {
+            List<Record> records = StreamSupport.stream(batch.spliterator(), false).toList();
+            if (records.isEmpty()) {
+                builder.writeBatch(batch);
+            }
+            else {
+                // TODO per-thread caching
+                ByteBuffer recordBuffer = ByteBuffer.allocate(recordBufferInitialBytes);
```

Comment:
If you mean the `IntFunction<ByteBufferOutputStream> bufferAllocator` which we pass to the encryption manager and which does the allocation via Netty then no, because that allocates `ByteBufferOutputStream` not `ByteBuffer`. To do the per-thread caching I think we'll need to add some new method(s) to the `FilterContext`. But I'm happy to kick this one to a later PR.

Reply:
>If you mean the `IntFunction<ByteBufferOutputStream> bufferAllocator` which we pass to the encryption manager and which does the allocation via Netty then no, because that allocates `ByteBufferOutputStream` not `ByteBuffer`

I did, but you make a good point about types. I wonder if we should be exposing the those new methods on the `FilterContext` sooner rather than later and then we can switch allocation strategies underneath them? As its not as simple as just calling the `buggerAllocator` I'm happy to defer the work.

Reply:
I eagerly await your PR :-)